### ISSUE TITLE
If I use torch.compile to compile the whole graph，in the my own compiler, how to manage the memory in my own compiler? 

### ISSUE THREAD

**Issue by mollon650** [pytorch/pytorch#117602]
### 🐛 Describe the bug

if I use torch.compile to compile the whole graph，in the my own compiler ，in forward stage，
1.if I enable memory reuse in the forward pass，how the backwards get the activation to calcute the gradient？has there some example in pytorch？
2.if i disable memory reuse，if i enable some op fusion，A op+B op fuison to one op, so the A output value is in sram or local memory or gloabal memory , torch can’t get the activation, in backwards how to calcute the gradient?has there some example in pytorch？
3.how to manage the memory in my own compiler to use torch.compile to speed up the training?
4.how the backwards (autogradient) get the activation from my own compiler ? the memory of every op in the graph must be in ddr?

### Error logs

none

### Minified repro

None

### Versions

None

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

**Comment by xmfan** [comment]
I think questions 1, 3 and 4 are asking about how gradients are calculated? If so, I would recommend reading through https://pytorch.org/docs/stable/notes/autograd.html, especially the [save_for_backward](https://pytorch.org/docs/stable/generated/torch.autograd.function.FunctionCtx.save_for_backward.html#torch.autograd.function.FunctionCtx.save_for_backward) API. For operations defined within PyTorch e.g. see `torch.pow` example in the doc, tensor saving is done automatically.

For question 3, you can defer graph handling to your own compiler by registering a custom backend for TorchDynamo to call. TorchDynamo will call your compiler with the captured graph. We have some examples of custom compiler backends in: https://pytorch.org/docs/stable/torch.compiler_custom_backends.html.

**Comment by mollon650** [comment]
I mean if i use fusion pass in compiler，Whether it is necessary to output the output of the fused intermediate operator to torch？for backwards to calculate the gradient

**Comment by bdhirsh** [comment]
@mollon650 you might want to take a look at https://dev-discuss.pytorch.org/t/min-cut-optimal-recomputation-i-e-activation-checkpointing-with-aotautograd/467.

When dealing with training, we eventually need to end up with a separate forward graph and backward graph to compile. In eager, the autograd engine will (by default) save intermediates for backward whenever they are needed in the backward graph. There are a couple of ways to enable memory re-use:

(1) activation checkpointing (user-controlled). The user can pretty specify what they want to be saved for the backward graph vs. recomputed in the backward. In this case, the compiler has no choice and should probably respect the user's choice for what to save vs. recompute

(2) torch.compile partitioning. If you are a custom backend that is integrating into `torch.compile` underneath AOTAutograd, then AOTAutograd will run a min-cut partitioning pass (linked above), that will basically try to figure out when it is "cheap" to recompute things in the backward.

> I mean if i use fusion pass in compiler，Whether it is necessary to output the output of the fused intermediate operator to torch？for backwards to calculate the gradient

If you are a backend sitting below autograd (so you've left control over which activations will be saved vs. recomputed to the autograd engine / AOTAutograd), then you'll basically be handed a separate forward graph and backward graph to compile. Any activations that were chosen to be saved for the backward (and not recomputed) will show up as **outputs** of the forward graph, which will basically force any backend compiler to materialize these activations in memory.

**Comment by mollonpj** [comment]
1.So I can don’t care about the memory of fusion op, the graph get from fx has already partitioned，I just do my own optimize in my compiler,others don’t care about?
2. Another question can dynamo trace collective ops? When will this feature be completed?

**Comment by bdhirsh** [comment]
These questions would probably be better for the pytorch forums: https://discuss.pytorch.org/. But:

> 1.So I can don’t care about the memory of fusion op, the graph get from fx has already partitioned，I just do my own optimize in my compiler,others don’t care about?

I'm not sure I fully understand the question, but if the question is "do I have to worry about the fusions that I create causing problems for saved activations", the answer is no (if autograd needs an activation saved for backward, it will force you to return is as an output in the forward graph)

> 2. Another question can dynamo trace collective ops? When will this feature be completed

There are "functional collectives" today that can be properly traced into a graph with `torch.compile`. Tracking issue here: https://github.com/pytorch/pytorch/issues/93173

**Comment by mollonpj** [comment]
@bdhirsh Thanks ，the answer from torch forums too slow，the issue of #93173 is not closed，do u know when it can finished ？which functional collectives can traced into graph?

**Comment by wconstab** [comment]
Dynamo supports tracing functional collectives, and can automatically convert some c10d collectives to functional ones. See https://github.com/pytorch/pytorch/blob/main/torch/distributed/_functional_collectives.py for more info.

**Comment by mollon650** [comment]
![image](https://github.com/pytorch/pytorch/assets/97667260/df47458b-d0ff-4e59-be80-01b099d0b166)
you means these c10d collectives just cvt to functional collectives，but it can't in fx graph ,it just run in eager mode.
so this problem can fixed in the later version? collectives can be in fx graph to compile the whole fx graph to improve the perf?

**Comment by xmfan** [comment]
Dynamo has some support to rewrite a c10d collective to a functional one, it should cover all_reduce.
https://github.com/pytorch/pytorch/blob/6e4e81a9ef6790d1593b9408e9ca7a2728a0af91/torch/distributed/_functional_collectives.py#L844-L848

Alternatively, you can also modify your script to directly use functional collectives in `torch.distributed._functional_collectives`.

If you're running into an issue, you can share a code snippet for us to look at.

**Comment by wconstab** [comment]
regarding the screenshot above, the dynamo_undupported_ops list is a little confusing.  We do support some forms of allreduce in the graph, but if async_op=True then we do cause a graphbreak and run in eager.

**Comment by mollonpj** [comment]
![IMG_4584](https://github.com/pytorch/pytorch/assets/5770287/115fffda-2b92-4fcd-97c2-e1338226b428)
![IMG_4583](https://github.com/pytorch/pytorch/assets/5770287/49cda9db-432f-4b91-af9d-957ef564054f)
1.So the dpp donot need to use tricks to improve the perf , it can use dynamo trace all reduce  and opt the whole graph make compute and all reduce overlap to improve perf?
2. when the other collectives （all gather reduce scatter broadcast ….）can traced in dynamo？

**Comment by wconstab** [comment]
See https://github.com/pytorch/pytorch/issues/109774 for more info on DDP tracing.  The allreduce op itself can already be traced, but bc the DDP engine is in C++ it isn't traced by dynamo.  We need compiled autograd to enable tracing the backwards part including the allreduce calls.

**Comment by wconstab** [comment]
i'm not sure what you mean by using your own compiler.  If you are using torch.compile with your own backend, then you're stuck with the same behavior as the rest of torch.compile.  But if you built a new compiler outside of torch.compile, i can't imagine what its behavior is going to be.

**Comment by xmfan** [comment]
I'm closing the issue, bug reports are meant to report errors when running code. If you have such a case, you can open a new issue and share the code that reproduces the bug.

For general questions on how PyTorch can be used, please use http://discuss.pytorch.org/ going forward.
